Redesign "Show Code" feedback

Hamms opened the pull request:
As per our [recent design work](https://drive.google.com/a/code.org/folderview?id=0B42L3wcu7t_gbTJWVlJROUI1R00&usp=sharing)

Also refactors both the feedback dialog "Show Code" and the standalone
"show code" button to use React

| Before | After |
| --- | --- |
| ![image](https://cloud.githubusercontent.com/assets/244100/13755766/ffefa9b6-e9d9-11e5-8c24-121df34c3ae4.png) | ![image](https://cloud.githubusercontent.com/assets/244100/13755233/779b78bc-e9d7-11e5-9040-01781b1ca582.png) |
| ![image](https://cloud.githubusercontent.com/assets/244100/13755776/090ff46a-e9da-11e5-90b6-7c0160fe495d.png) | ![image](https://cloud.githubusercontent.com/assets/244100/13755333/fbbff2f8-e9d7-11e5-9dc2-de0cb1207ec4.png) |

Bjvanminnen commented on apps/src/feedback.js:
```diff
@@ -807,19 +810,18 @@ FeedbackUtils.prototype.getShowCodeElement_ = function(options) {
+  ReactDOM.render(<CodeWritten
+    numLinesWritten = {numLinesWritten}
+    totalNumLinesWritten = {totalNumLinesWritten}
+    generatedCodeElement = {generatedCodeElement}
```
nit: In JSX, I never really see spaces surround the equals sign. I suspect this is because it's supposed to look HTML like. (i.e. I would expect `numLinesWritten={numLinesWritten}`

Bjvanminnen replied:
Would it make more sense to express this as:

```
<CodeWritten ...>
  {generatedCodeComponent}
</CodeWritten>
```

i.e. passing in the component as a child instead of a prop?

Hamms replied:
Oh shoot that's way better. I hadn't considered that as a new option now that we can inline JSX!

Bjvanminnen commented on apps/src/feedback.js:
```diff
@@ -883,13 +886,7 @@ FeedbackUtils.prototype.getGeneratedCodeElement_ = function(options) {
-
-  return codeDiv;
+  return (<Code message={infoMessage} code={code} />);
 };
```
Actually, I wonder if it would make the most sense to have this method return our message and code, and then have the caller create the React component. I think it will be less confusing if we limit the number of places we're passing around React components.

Hamms replied:
Ooooh, can we use the [JSX Spread operator](https://facebook.github.io/react/docs/jsx-spread.html) ?

``` javascript
var generatedCodeProperties = this.getGeneratedCodeProperties_(options);

return (<Code {...generatedCodeProperties} />);
```

Bjvanminnen replied:
We can, though my own preference would to mostly reserve that for cases where we have more dynamic properties. In this case, I feel like I'd rather not have to go peek into getGeneratedCodeProperties to know what I was setting on Code. In a world with ES6 (actually, technically this might need ES7 for object destructing) we could do:

```
var { message, code } = this.getGeneratedCodeProperites_(options);
return <Code message={message} code={code}/>;
```

Bjvanminnen commented on apps/src/feedback.js:
```diff
@@ -919,24 +916,24 @@ FeedbackUtils.prototype.getGeneratedCodeDescription = function (codeInfoMsgParam
-  }), buttons);
-  codeDiv.appendChild(buttons);
+  ReactDOM.render(<ShowCode
+    generatedCodeComponent={generatedCodeComponent}
```
Same suggestion that we maybe pass this in as a child instead of a prop. (The component can then access it as `this.props.children`)

Hamms replied:
Actually, this component is extremely simple. How would we feel about inlining the whole thing?

``` javascript
ReactDOM.render(<div>
  {generatedCodeComponent}
  <DialogButtons ok={true} />
</div>, codeDiv);
```

Bjvanminnen replied:
Seems fine. Especially if this actually becomes:

```
<div>
  <Code message={geneatedCodeInfo.message} code={generatedCodeInfo.code}/>
  <DialogButtons ok={true}/>
</div>
```

Bjvanminnen commented on apps/src/templates/feedback/Code.jsx:
```diff
@@ -0,0 +1,17 @@
+    return (<div className="generated-code-container">
+      <p className="generatedCodeMessage" dangerouslySetInnerHTML={{__html: this.props.message}} />
+      <pre className="generatedCode" dangerouslySetInnerHTML={{ __html: this.props.code }} />
+    </div>);
```
I'm wondering at what point we should start worrying about how frequently we're using dangerouslySetInnerHTML. It's starting to sound a lot less dangerous to me than it once did as it becomes common place. What do you think?

Bjvanminnen replied:
Can we replace these classNames with inlined styles? It looks like generatedCodeMessage does have some hover styling, but maybe that's insignificant enough that we don't mind losing it?

FWIW, I've explored adding radium, which would allow us to do hover for inlined styles. The main reason I'm somewhat wary to add it is it does increase our file size some (an extra 50k or so I think)

Hamms replied:
It seems like the reason React is so paranoid about rendering unescaped HTML is that it assumes you're doing this primarily in situations where you're rendering user-created content, and that in all other situations you should just render another React component.

It seems like we primarily use dangerouslySetInnerHTML in situations where we want to render pre-rendered content that had been generated by some other non-react part of our system, most often stuff that comes from levelbuilder. It's sort of a necessary evil as long as our system is not 100% React.

(note in this case, the `this.props.code` element actually IS rendering user-created content, albiet some heavily processed and controlled content).

We can replace `generatedCode` with an inline style, but `generatedCodeMessage` is using the class to style some of that `dangerouslySetInnerHTML` content.

Bjvanminnen commented on apps/style/common.scss:
```diff
@@ -1201,6 +1201,22 @@ $workspace-header-button-margin: ($workspace-headers-height - $workspace-header-
+    margin: 25px 10px 25px -80px;
+  }
+}
+
```
Can we do these as inline styles in the React component?

Hamms replied:
One issue with inline styles (both for this and for the other Code component) is that it makes it difficult to situationally override styles. In particular, the minecraft styles have a couple overrides apiece for all of these components.

Bjvanminnen replied:
One option would be to pass in isMinecraft as a prop and style it differently depending on the value. Another would be to have a default set of styles, and allow the caller to pass in override styles.

If this is going to be too big a PITA, we can leave as is, but I really would like to get away from having styles living externally to the components. It makes it a lot harder to track down what the style is.

Hamms replied:
it's less a PITA, and more a situation where I'm not convinced that the advantages of tightly coupled style+layout outweigh the advantages of loosely coupled style+layout; losing things like our `$color` and `$font` variables (which I'm not using here but really should be) and an already-existing framework for situational overrides.

On the other hand, if we keep some styles for our React stuff in CSS and use inline styles for some other react stuff, then it becomes even harder to track down what the style is.

Bjvanminnen replied:
For stuff like colors, I've started moving some of them into sharedJsxStyles.js. Font could probably go in here as well, tho it does mean we have one copy in css and another in js.

Bjvanminnen commented:
Done with first pass of this PR. Generally looks good, though I had some suggestions on ways we might restructure some things. Let me know if you disagree with anything.

Hamms commented:
I'm super torn on whether or not it would be best to inline these styles. Or even if it would be best to inline some of them.

I get the sense that React's inline styles work awesome when used as a replacement for stateful CSS classes, but inlining larger-scope styling seems inappropriate. We definitely have a problem with our CSS being unorganized and slightly scattered, but this seems like a less-than-ideal way to go about addressing that problem.

Do we have any kind of org-wide Best Practices for React yet? Even if we do, I assume we don't yet have one for styling React.

Bjvanminnen commented:
We agreed as a team that our goal was to do inline styles where we could. We never really went into more detail than that.

Hamms commented:
I moved some styles around, but left in CSS some styles that need to be overridden. Also fixed a bug that was preventing tests from running.

Did you want to give this another quick look?

Bjvanminnen commented on apps/src/templates/feedback/CodeWritten.jsx:
```diff
@@ -0,0 +1,46 @@
+        lineHeight: '20px',
+        fontWeight: 'normal',
+        outline: 'none',
+        padding: '5px',
```
Note: If you want, you can also just do `fontSize: 18` and React will automatically append 'px'.

Bjvanminnen commented:
lgtm

Hamms commented:
Thanks a bunch for all your feedback!
